documentation/writing_tests.md

```diff
@@ -0,0 +1,1534 @@
+
+```ts
+// Our commentary on arrow functions for test cases
+// does not apply outside of the declaration of test cases.
+// You can declare arrow functions anywhere comfortably,
+// just not where `it()` is used.
+
+describe("My async tests", () => {
+  const myAsyncMethod = async () => {
+    return true;
+  }
+
+  it.skip("Discouraged async test with callback", function(done) {
+    // Do not do this. Use async/await
+    myAsyncMethod.then(done);
+  });
+
+  it.skip("Discouraged async test with arrow async function", async () => {
+    // Do not use arrow functions to declare test cases,
+    // tests will lose Mocha's context, which is useful, and needed by the Recorder.
+    await myAsyncMethod();
+  });
+
+  it("Recommended async test with natural async function", async function() {
+    // This is good ✅
+    await myAsyncMethod();
+  });
+});
+```
+
+### Chai
+
+Chai is an assertion library, similar to Node's built-in `assert`. While `assert` is a module that belongs to NodeJS's standard library, it was mainly designed to write tests for the NodeJS's core, and [it has been discouraged from being used as a dependency](https://github.com/nodejs/node/issues/4532). Chai is very popular, and it is widely used for both NodeJS and the browser. You can read more about Chai through their main website: https://www.chaijs.com/
+
+Chai makes testing much easier by providing an extensive list of assertions that can run against your code. This list of assertions can be seen in detail by going to their assertion style guide: <https://www.chaijs.com/guide/styles/>
+
+Another important learning resource for Chai is: Chai Assertions for Promises <https://www.chaijs.com/plugins/chai-as-promised/>.
+
+#### Recommended Chai assertion style
+
+The Azure SDK for JavaScript (and TypeScript) prefers Chai's Assert style. For example:
+
+```ts
+import { assert } from "chai";
+
+let foo = "bar";
+let beverages = { tea: [ 'chai', 'matcha', 'oolong' ] };
+
+assert.typeOf(foo, "string");
+assert.typeOf(foo, "string", "foo is a string");
+assert.equal(foo, "bar", "foo equals to `bar`");
+assert.lengthOf(foo, 3, 'foo`s value has a length of 3');
+assert.lengthOf(beverages.tea, 3, 'beverages has 3 types of tea');
+```
+
+#### Chai in our dependencies
+
+Since we're using [Rush](https://rushjs.io/), we have to use the same version of our packages in each one of the projects inside of this repository. If you want to add `chai` as a dev dependency to a new project inside of this repository, first make sure you are in the root folder of that project, then you can run the following command:
+
+```
+rush add --dev -p chai
+```
+
+Since we're using TypeScript, make sure to also add the types:
+
+```
+rush add --dev -p @types/chai
+```
+
+### Rollup
```

Comment:
When you say to separate this, do you mean in several documents? I can make a `test_guidelines` folder and split this one in several files, and link them through the `index.md`. Would that help?

Reply:
I think *maybe* it would help, since this document is pretty big and that can spook people. 

On the other hand, it's kinda nice that it's one document and searchable, so I don't feel strongly about this.

Reply:
I'll leave this open for debate!